src/sim/detail/InitializeTracks.cu

```diff
@@ -43,56 +43,54 @@ struct IsEqual
  * from either primary particles or secondaries. The new tracks are inserted
  * into empty slots (vacancies) in the track vector.
  */
-__global__ void init_tracks_kernel(const StatePointers            states,
-                                   const ParamPointers            params,
-                                   const TrackInitializerPointers inits,
-                                   size_type num_vacancies)
+__global__ void init_tracks_kernel(const ParamsDeviceRef         params,
+                                   const StateDeviceRef          states,
+                                   const TrackInitStateDeviceRef inits,
+                                   size_type                     num_vacancies)
 {
-    auto thread_id = KernelParamCalculator::thread_id().get();
-    if (thread_id < num_vacancies)
+    auto tid = KernelParamCalculator::thread_id();
+    if (!(tid < num_vacancies))
+        return;
+
+    // Get the track initializer from the back of the vector. Since new
+    // initializers are pushed to the back of the vector, these will be the
+    // most recently added and therefore the ones that still might have a
+    // parent they can copy the geometry state from.
+    const TrackInitializer& init
+        = inits.initializers[ThreadId(inits.initializers.size() - tid.get() - 1)];
```

Comment:
Since this uses unsigned integer arithmetic which can overflow, could you check this by inserting:
```
CELER_ASSERT(tid.get() + 1 <= inits.initializers.size());
```
and same right before `vac_id`?